separatedunit.hpp

```diff
@@ -0,0 +1,185 @@
+    void RTDyad(int numerator, int denominator, bitLenInt qubitIndex);
+    void RX(double radians, bitLenInt qubitIndex);
+    void RXDyad(int numerator, int denominator, bitLenInt qubitIndex);
+    void RY(double radians, bitLenInt qubitIndex);
+    void RYDyad(int numerator, int denominator, bitLenInt qubitIndex);
+    void RZ(double radians, bitLenInt qubitIndex);
+    void RZDyad(int numerator, int denominator, bitLenInt qubitIndex);
+
+    void CRT(double radians, bitLenInt control, bitLenInt target);
+    void CRTDyad(int numerator, int denominator, bitLenInt control, bitLenInt target);
+    void CRY(double radians, bitLenInt control, bitLenInt target);
+    void CRYDyad(int numerator, int denominator, bitLenInt control, bitLenInt target);
+    void CRZ(double radians, bitLenInt control, bitLenInt target);
+    void CRZDyad(int numerator, int denominator, bitLenInt control, bitLenInt target);
+
+    void ROL(bitLenInt shift, bitLenInt start, bitLenInt length);
+    void ROR(bitLenInt shift, bitLenInt start, bitLenInt length);
+
+    void INC(bitCapInt toAdd, bitLenInt start, bitLenInt length);
+    void INCC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt carryIndex);
+    void INCS(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt overflowIndex);
+    void INCSC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt overflowIndex, bitLenInt carryIndex);
+    void INCSC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt carryIndex);
+    void INCBCD(bitCapInt toAdd, bitLenInt start, bitLenInt length);
+    void INCBCDC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt carryIndex);
+    void DEC(bitCapInt toSub, bitLenInt start, bitLenInt length);
+    void DECC(bitCapInt toSub, bitLenInt start, bitLenInt length, bitLenInt carryIndex);
+    void DECS(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt overflowIndex);
+    void DECSC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt overflowIndex, bitLenInt carryIndex);
+    void DECSC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt carryIndex);
+    void DECBCD(bitCapInt toAdd, bitLenInt start, bitLenInt length);
+    void DECBCDC(bitCapInt toSub, bitLenInt start, bitLenInt length, bitLenInt carryIndex);
+
+    void QFT(bitLenInt start, bitLenInt length);
+    void ZeroPhaseFlip(bitLenInt start, bitLenInt length);
+    void CPhaseFlipIfLess(bitCapInt greaterPerm, bitLenInt start, bitLenInt length, bitLenInt flagIndex);
+    void PhaseFlip();
+
+    unsigned char SuperposeReg8(bitLenInt inputStart, bitLenInt outputStart, unsigned char* values);
+    unsigned char AdcSuperposeReg8(
+        bitLenInt inputStart, bitLenInt outputStart, bitLenInt carryIndex, unsigned char* values);
+    unsigned char SbcSuperposeReg8(
+        bitLenInt inputStart, bitLenInt outputStart, bitLenInt carryIndex, unsigned char* values);
+
+protected:
+    std::unique_ptr<QbLookup[]> qubitLookup;
```

Comment:
95% sure you want a vector here, and for `qubitInverseLookup`.  But I haven't gotten to the rest of the code, yet.

Reply:
Just to emphasize: broadly speaking, wherever you were allocating an array, it’s generally better to use a vector instead - that way you don’t have to worry about memory management, the access semantics are the same, etc.  Means you don’t need the unique_ptr either.